File: rb/algorithms/lib/sorting/quick.rb
require_relative 'sorting_helpers'

module Sorting
  class Quick
    include SortingHelpers

    def sort!(collection)
      collection.shuffle!
      do_sort!(collection, 0, (collection.length - 1))
    end

    private

    def do_sort!(collection, lo, hi)
      return if hi <= lo
      j = partition(collection, lo, hi)
      do_sort!(collection, lo, (j - 1))
      do_sort!(collection, (j + 1), hi)
    end

    def partition(collection, lo, hi)
      i, j = lo, (hi + 1)
      v = collection[lo]

      loop do
        i += 1

        while collection[i] < v
          break if i == hi
          i = i.succ
        end

        j -= 1

        while v < collection[j]
          break if j == lo
          j -= 1
        end

        break if i >= j

        swap!(collection, i, j)
      end

      swap!(collection, lo, j)
      j
    end
  end
end

File: lib/sorting/quick.js
'use strict';

const shuffle = require('shuffle-array');
const swap = require('../util').swap;

function partition(ary, lo, hi) {
  const x = ary[hi];
  let i = lo - 1;

  for (let j = lo; j <= (hi - 1); j++) {
    if (ary[j] <= x) {
      i++;
      swap(ary, i, j);
    }
  }

  swap(ary, i + 1, hi);
  return i + 1;
}

function do_sort(ary, lo, hi) {
  if (hi <= lo) return;
  const j = partition(ary, lo, hi);
  do_sort(ary, lo, j - 1);
  do_sort(ary, j + 1, hi);
}

module.exports = (ary) => {
  shuffle(ary);
  do_sort(ary, 0, ary.length - 1);
  return ary;
}
